### PR TITLE
Scan untracked files in git for packaging

### LINKED ISSUE #797: `cargo package` is not working
The `cargo package` command results in:

```
   Packaging hello_world v0.0.1 (file:///Users/eduardobautista/Projects/hello_world)
   Verifying hello_world v0.0.1 (file:///Users/eduardobautista/Projects/hello_world)
failed to verify package tarball
```

This is on a newly generated cargo project using the October 31st build `492bc86` on OS X.

### PR DESCRIPTION
**Pull request by alexcrichton**
Closes #797

### CODE REVIEW

**Comment by huonw** [comment]
Hm, I'm not sure that we want to implicitly add files just lying around? To me it seems similar to https://github.com/rust-lang/cargo/issues/880.

Maybe we could special-case the "you have no commits" case, or do the `git add .` thing, but implicitly putting things on the internet makes me nervous (I know you're *meant* to ignore files containing private keys etc. but it's a mistake waiting to happen).

**Comment by alexcrichton** [comment]
That's not a bad idea, I've updated to only use `git status` if there is no HEAD

File: src/cargo/sources/path.rs [hunk 1]
```diff
@@ -113,23 +113,36 @@ impl PathSource {
...
+                                .filter(|_| unborn);
+
+        'outer: for file_path in index_files.chain(untracked) {
             // Filter out files outside this package.
             if !pkg_path.is_ancestor_of(&file_path) { continue }
 
             // Filter out Cargo.lock and target always
-            if fname == b"Cargo.lock" { continue }
-            if fname == b"target" { continue }
+            if file_path.filename() == Some(b"Cargo.lock") { continue }
```
Comment (huonw) [inline src/cargo/sources/path.rs]: (Incidental; is crates.io meant to support binary packages?)

**Comment by huonw** [comment]
Hm, I'm still nervous about this it feels a lot like the software doesn't have enough info to truly "do what I mean" properly the case that there's nothing committed. It's easy to just publish a bunch of stuff unintentionally, I would like to see some output from cargo indicating to the user that something possibly unexpected is happening, or even just emitting an error "cargo does not have enough information to decide which files to include, considering making a commit".

(I think https://github.com/rust-lang/cargo/issues/883 would assist with this too.)

**Comment by huonw** [comment]
(I suppose having a hard error only makes sense if there's a way to explicitly whitelist files, so that people can e.g. use alternative VCS's.)

**Comment by alexcrichton** [comment]
I think I'll try to tackle this later, got a lot of other stuff going on now!

**Pull request closed without merging**